app/modules/zoo-api/zooapi.factory.js

```diff
@@ -4,10 +4,21 @@ require('./zooapi.module.js')
     .factory('zooAPI', zooAPI);
 
 var ApiClient = require('panoptes-client/lib/api-client');
-
+var OAuth = require('panoptes-client/lib/oauth')
 // @ngInject
-function zooAPI(zooAPIConfig) {
-
+function zooAPI($rootScope, zooAPIConfig) {
+    ApiClient.beforeEveryRequest = function() {
+        return OAuth.checkBearerToken()
+            .then(function (token) {
+                console.log('Token refreshed: ', token);
+            })
+            .catch(function (error) {
+                console.log('Failed to refresh token: ', error);
+                alert('Your session has finished. Please save your work and login again.')
```

Comment:
How does this alert allow them to save their work? My reading is this will tell them to save their work and the press ok/cancel and then it'll sign them out anyway.

Reply:
You're right. Building on Jim's suggestion, I think a better flow would be to save work on clicking `OK` in the alert and then call `OAuth.signIn(url)` to start a new session.

Reply:
If your alert tells the user that "pressing ok will get them a new session and reload their work" then it should be a couple of page loads and they'll be back where they left...right? That seems reasonable to me if i press OK. 

This is a matter of signalling what's going to happen to the user and letting them decide.

Reply:
Yep, I was thinking something like "Your session is about to expire. Press ok to save your work and get a new session."